pipeline/download_gene.py

```diff
@@ -0,0 +1,154 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script that download and postprocess ISH dataset from Allen Brain."""
+from __future__ import annotations
+
+import argparse
+import logging
+import sys
+from pathlib import Path
+
+import numpy as np
+import PIL
+
+
+logger = logging.getLogger("download-gene")
+
+
+def parse_args():
+    """Parse command line arguments.
+
+    Returns
+    -------
+    args : argparse.Namespace
+        The parsed command line arguments.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--gene-name",
+        type=str,
+        help="""\
+        Path to CCFv3 annotation volume.
+        """,
+    )
+    parser.add_argument(
+        "--output-dir",
+        type=Path,
+        help="""\
+        Path to output directory to save results.
+        """,
+    )
+    args = parser.parse_args()
+
+    return args
+
+
+def postprocess_dataset(dataset):
+    """Post process given dataset.
+
+    Parameters
+    ----------
+    dataset : iterable
+        List containing dataset, each element being tuple
+        (img_id, section_coordinate, img, df)
+
+    Returns
+    -------
+    dataset_np : np.ndarray
+        Array containing gene expressions of the dataset.
+    expression_np : np.ndarray
+        Array containing image expression of the dataset.
+    metadata_dict : dict
+        Dictionary containing metadata of the dataset.
+        Keys are section numbers and image ids.
+    """
+    metadata_dict = {}
+    section_numbers = []
+    image_ids = []
+    expression_np = []
+    dataset_np = []
+
+    for img_id, section_coordinate, img, img_expression, df in dataset:
+        if section_coordinate is None:
+            # In `atldld.sync.download_dataset` if there is a problem during the download
+            # the generator returns `(img_id, None, None, None, None)
+            # TODO: maybe notify the user somehow?
+            continue
+
+        section_numbers.append(section_coordinate // 25)
+        image_ids.append(img_id)
+        warped_img = 255 - df.warp(img, border_mode="constant", c=img[0, 0, :].tolist())
+        warped_exp = df.warp(
+            img_expression, border_mode="constant", c=img[255, 255, :].tolist()
+        )
+        dataset_np.append(warped_img)
+        expression_np.append(warped_exp)
+
+    dataset_np = np.array(dataset_np)
+    expression_np = np.array(expression_np)
+
+    metadata_dict["section_numbers"] = section_numbers
+    metadata_dict["image_ids"] = image_ids
+    metadata_dict["image_shape"] = warped_img.shape
```

Comment:
The shape after synchronization should always be identical, right? Just wonder whether this piece of metadata is useful.

Reply:
The shape can be different if we go for different resolutions (and it might be the case in the future). It also helps to know if it is coronal or sagittal. However, I am not sure we use this directly for now, so maybe we can remove it.

Reply:
Make sense:) We can keep it there:)